nebu/tests/cli/test_cnxml_to_html.py

```diff
@@ -0,0 +1,63 @@
+import os
+from unittest import mock
+
+import pytest
+
+
+def get_files(directory, filename):
+    for root, dirs, files in os.walk(str(directory)):
+        if filename in files:
+            yield os.path.join(root, filename)
+
+
+@pytest.fixture
+def clear_files(request):
+    def _clear_files(directory, filename):
+        def cleanup():
+            for f in get_files(directory, filename):
+                os.unlink(f)
+
+        # make sure files are removed before tests
+        cleanup()
+
+        # make sure files are removed after tests
+        request.addfinalizer(cleanup)
+
+    return _clear_files
+
+
+@pytest.fixture
+def cnxml_to_full_html(request):
+    patcher = mock.patch('nebu.cli.cnxml_to_html.cnxml_to_full_html')
+    mock_function = patcher.start()
+    mock_function.return_value = '<html><body>transformed html</body></html>'
+    request.addfinalizer(patcher.stop)
+    return mock_function
+
+
+class TestCnxmlToHtmlCmd:
+    def test(self, datadir, invoker, clear_files, cnxml_to_full_html):
+        col_path = datadir / 'collection_no_changes'
+
+        # make sure index.cnxml.html doesn't exist
+        clear_files(col_path, 'index.cnxml.html')
+        index_cnxml_html = list(get_files(col_path, 'index.cnxml.html'))
+        assert index_cnxml_html == []
+
+        from nebu.cli.main import cli
+        args = ['cnxml-to-html', str(col_path)]
+        result = invoker(cli, args)
+        assert result.exit_code == 0
+
+        # make sure the number of index.cnxml.html is the same as index.cnxml
+        index_cnxml = list(get_files(col_path, 'index.cnxml'))
+        index_cnxml_html = list(get_files(col_path, 'index.cnxml.html'))
+        assert len(index_cnxml) == len(index_cnxml_html)
+        assert cnxml_to_full_html.call_count == len(index_cnxml)
+
+        # check that the index.cnxml.html is indeed transformed with the mocked
+        # cnxml_to_full_html function
+        for path in index_cnxml_html:
+            with open(path) as f:
+                content = f.read()
+            assert '<html><body>transformed html</body></html>' == content
```

Comment:
Isn't this just testing that the mocking mechanism worked? When you mocked `cnxml_to_full_html` you also set its return value and that's all this line is testing. In Rails, for example, I would've called `.and_call_original` to actually test that the function works, not just that it gets called as many times as we expect (which is good, btw, and thanks for adding that after I requested, but not sufficient). In other words, it seems to me like you're breaking rule # 1 of mocking: "Don't mock what you don't own." @karenc

Reply:
I thought that's what you wanted, that `cnxml_to_full_html` was indeed called.  I have thought about calling the original function.  To me, that's the difference between unit testing and integration testing.  By mocking `cnxml_to_full_html`, I'm doing unit testing.  This only tests the code that I wrote in neb and not outside of it.  It works as long as `cnxml_to_full_html` doesn't change its interface.

Reply:
> To me, that's the difference between unit testing and integration testing

True. I guess I expected one unit test and one integration test. The unit test just tests your code and it wouldn't break if `cnxml_to_full_html` breaks or changes 👍, but the integration test would (break), as it should, and then we'd have to fix just that one test. But I suppose that does sound like an improvement or a bug fix, so this one deserves/deserved to get merged and I should open up a separate issue for adding an integration test. No problem.